File: assets/static/blog/dsa/data_structures/binary_heap/implementation.py

class BinaryHeap:
    def __init__(self):
        self._heap = [0]
        self.size = 0

    def __repr__(self):
        return str(self._heap)

    def _perc_up(self):
        """Percolate a newly inserted node at the bottom of the tree upwards until the node is
        less than or equal to both of its children
        """

        # Start with the bottom-right node
        node_ind = self.size
        parent_ind = node_ind // 2

        while parent_ind > 0:
            # If the parent node is less than the newly inserted node, swap them and continue.
            parent_ind = node_ind // 2
            if self._heap[node_ind] < self._heap[parent_ind]:
                self._heap[parent_ind], self._heap[node_ind] = self._heap[node_ind], self._heap[parent_ind]
            node_ind = parent_ind

    def _perc_down(self, node_ind):
        """Percolate a given node down until it's less than or equal to both of its children"""

        # If there is no left child node, we're at the bottom of the tree and there's nothing to do
        while (node_ind * 2) <= self.size:

            # If one of the childeren nodes is less than the given node, swap them and continue
            min_child_ind = self._min_child(node_ind)
            if self._heap[node_ind] > self._heap[min_child_ind]:
                self._heap[node_ind], self._heap[min_child_ind] = self._heap[min_child_ind], self._heap[node_ind]
            node_ind = min_child_ind



    def _min_child(self, node_ind):
        """Return the index of the minimum child node for a given node index"""
        left_node_ind = node_ind * 2
        right_node_ind = left_node_ind + 1

        if right_node_ind > self.size:
            return left_node_ind

        if self._heap[left_node_ind] < self._heap[right_node_ind]:
            return left_node_ind
        return right_node_ind

    def insert(self, val):
        """Insert a new node at the bottom-right of the tree, and then percolate the new node up if needed"""
        self._heap.append(val)
        self.size += 1
        self._perc_up()

    def pop_min(self):
        min_node = self._heap[1]
        self._heap[1] = self._heap[self.size]
        self.size -= 1
        self._heap.pop()
        self._perc_down(1)
        return min_node

    def build_heap(self, list):
        self.size = len(list)
        node_ind = self.size // 2
        self._heap = [0] + list
        while node_ind > 0:
            self._perc_down(node_ind)
            node_ind -= 1



inp = [1,10,5,6,11,22,3,50,1,2,51]

a = BinaryHeap()
a.build_heap(inp)

b = BinaryHeap()
for val in inp:
    b.insert(val)

assert a._heap == b._heap

File: assets/static/blog/dsa/data_structures/directed_graph/implementation.go
package main
import "fmt"


type BinaryHeap struct {
  data []int
  size int
}


// Percolate a newly inserted node up until it's less than or equal to both of its children
func (heap *BinaryHeap) percUp() {

  new_node_ind := heap.size - 1
  parent_node_ind := new_node_ind / 2

  for (heap.data[new_node_ind] < heap.data[parent_node_ind]) {
    heap.data[new_node_ind], heap.data[parent_node_ind] = heap.data[parent_node_ind], heap.data[new_node_ind]
    new_node_ind = parent_node_ind
    parent_node_ind = new_node_ind / 2
  }
}

// Percolate a given node down until it's less than or equal to its children
func (heap *BinaryHeap) percDown(node_ind int) {

  min_child_ind := heap.minChild(node_ind)
  for heap.data[node_ind] > heap.data[min_child_ind] {
      heap.data[min_child_ind], heap.data[node_ind] = heap.data[node_ind], heap.data[min_child_ind]
      node_ind = min_child_ind
      min_child_ind = heap.minChild(node_ind)
  }
}

// Return the index of the minimum node between right and left childeren of a given node
func (heap *BinaryHeap) minChild(node_ind int) int {
  left_child_ind := node_ind * 2
  right_child_ind := left_child_ind + 1

  if node_ind == 0 {
    left_child_ind += 1
    right_child_ind += 1
  }

  if left_child_ind >= heap.size {
    return node_ind
  }
  if heap.data[left_child_ind] < heap.data[right_child_ind] {
    return left_child_ind
  }
  return right_child_ind
}

// Insert a node at the bottom right of the tree and perc up until the tree is a heap
func (heap *BinaryHeap) Insert(val int) {
  heap.data = append(heap.data, val)
  heap.size += 1
  heap.percUp()
}

// Remove and return the root of the tree, replacing it with the bottom-right most node. Perc
// the new root down until the tree is a heap again
func (heap *BinaryHeap) PopMin() int {
  min_node := heap.data[0]
  heap.data[0] = heap.data[heap.size-1]
  heap.percDown(0)
  heap.data = heap.data[:heap.size-1]
  heap.size -= 1
  return min_node

}

// Create a new binary heap from a slice
func (heap *BinaryHeap) New(arr []int) {
  heap.data = arr

  i := heap.size - 1
  for i >= 0 {
    heap.percDown(i)
    i -= 1
  }
}


func main() {
  b := BinaryHeap{}
  b.Insert(2)
  b.Insert(3)
  b.Insert(1)
  b.Insert(4)
  b.Insert(40)
  b.Insert(4)
  b.Insert(3)
  b.Insert(0)


  fmt.Println(b.data)
  fmt.Println(b.PopMin())
  fmt.Println(b.data)

  t := []int{2,3,1,4,40,4,3,0}
  b.New(t)
  fmt.Println(b.data)

}
